File: axion-core/forge/rnc_engine.py
"""
[CORE] [ENGINE] [RNC_ENGINE]
Artifact ID: CORE.Engine.RNCEngine
Official Name: rnc_engine.py
Version: v15.0 [OMEGA]
Status: [CANONIZED]
Description: Validates RNC IDs and maps them to canonical filesystem paths.

[UIP-V15]
| Key | Value |
| :--- | :--- |
| **Artifact ID** | `CORE.Engine.RNCEngine` |
| **Official Name** | `rnc_engine.py` |
| **Version** | **v15.0 [OMEGA]** |
| **Domain** | `CORE` |
| **Status** | `[CANONIZED]` |
| **Relations** | `GOVERNED_BY: CORE-CODEX-001` |
"""

import os
import re
from typing import Dict, List

try:
    from .enums import Domain, SubSystem
except (ImportError, ValueError):
    from enum import Domain, SubSystem


class RNCEngine:
    """
    Sovereign Logic Engine for the Relational Naming Convention (RNC).
    """

    # [DOMAIN].[Subsystem].[Descriptor]
    RNC_PATTERN = re.compile(r"^([A-Z0-9]+)\.([A-Z0-9]+)\.([A-Za-z0-9_-]+)$")

    @classmethod
    def validate_id(cls, artifact_id: str) -> bool:
        """Verifies if an ID aligns with OMEGA v15.0 RNC standards."""
        match = cls.RNC_PATTERN.match(artifact_id)
        if not match:
            return False

        domain_str, subsystem_str, _ = match.groups()

        # Cross-reference with enums
        try:
            Domain(domain_str)
            SubSystem(subsystem_str)
            return True
        except ValueError:
            return False

    @classmethod
    def suggest_path(cls, artifact_id: str) -> str:
        """Maps an RNC ID to its canonical filesystem location."""
        match = cls.RNC_PATTERN.match(artifact_id)
        if not match:
            raise ValueError(f"Invalid RNC ID for path mapping: {artifact_id}")

        domain, subsystem, descriptor = match.groups()

        # Domain-to-Root Mapping
        root_map = {
            "GVRN": "_governance",
            "CORE": "axion-core",
            "LAB": "nova_forge",
            "SELT": "_governance/templates",
            "COMM": "03_Avatar",
        }

        root = root_map.get(domain, "unknown")

        # Subsystem-to-Folder Mapping (Example logic)
        # In a full implementation, this uses a refined lookup table
        folder_map = {
            "AVATAR": "03_Avatar",
            "REGISTRY": "01_Registries",
            "ENGINE": "forge",
            "LEARNING": "06_Learning",
        }

        folder = folder_map.get(subsystem, subsystem.lower())

        return os.path.join(root, folder, f"{artifact_id}.md")

    @classmethod
    def safe_transform(cls, file_path: str, transformer_func) -> None:
        """
        [SOVEREIGN SHIELD]
        Applies a programmatic transformation with state-buffering to prevent truncation.
        """
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"Cannot transform non-existent file: {file_path}")

        # 1. Read entire file into buffer (Kinetic -> Mind)
        with open(file_path, "r", encoding="utf-8") as f:
            content = f.read()

        # 2. Apply transformation in-memory
        transformed_content = transformer_func(content)

        # 3. Validation: Ensure we haven't nuked the file (Zero Entropy protection)
        if len(content) > 0 and len(transformed_content) == 0:
            raise RuntimeError(
                f"Safety Trigger: Transformation returned 0 bytes for {file_path}. Aborting write."
            )

        # 4. Atomic Write (Mind -> Substrate)
        with open(file_path, "w", encoding="utf-8", newline="\n") as f:
            f.write(transformed_content)

    @classmethod
    def sync_folder(cls, folder_path: str) -> List[Dict]:
        """Audits a directory for RNC compliance and returns a dissonance report."""
        report = []
        for root, _, files in os.walk(folder_path):
            for file in files:
                if file.endswith(".md"):
                    # Basic extraction of ID from filename or header
                    # (Simplified for the engine logic)
                    artifact_id = file.replace(".md", "")
                    if not cls.validate_id(artifact_id):
                        report.append(
                            {
                                "file": os.path.join(root, file),
                                "issue": "Non-compliant RNC ID",
                                "detected_id": artifact_id,
                            }
                        )
        return report


# [TERMINAL SIGNATURE]
# [OMNI-ARTIFACT-ANCHOR] ID: CORE.Engine.RNCEngine VER: v15.0 [OMEGA] STATUS: CANONIZED TS: 2026-04-23

File: nova_forge/core/engines/RNCEngine.ts
/*
artifact_anchor:
  id: COG.RNCENGINE.001
  version: v15.0 [OMEGA]
  provenance: '2026-07-22'
  domain: COG
  celestial_class: PLANET
  tier: LOGIC
  state: ACTIVE
  ethos: SOVEREIGN_LOGIC_COMPONENT
  relations: []
*/

import * as fs from "fs";
import * as path from "path";
import { PhoenixLogger } from "@system/logging";
import * as standards from "@domain/13_Standardization/GVRN.Standards.json";

export class RNCEngine {
  // [DOMAIN].[Subsystem].[Descriptor]
  private static RNC_PATTERN = /^([A-Z0-9]+)\.([A-Z0-9]+)\.([A-Za-z0-9_\-\.]+)$/;

  /**
   * Verifies if an ID aligns with OMEGA v15.0 RNC standards.
   */
  public static validateId(artifactId: string): boolean {
    const match = this.RNC_PATTERN.exec(artifactId);
    if (!match) {
      return false;
    }

    const domainStr = match[1];
    const subsystemStr = match[2];

    const validDomains = standards.enums.Domain;
    const validSubSystems = standards.enums.SubSystem;

    // Check if domain and subsystem exist in valid standards
    const hasDomain = Object.prototype.hasOwnProperty.call(validDomains, domainStr);
    const hasSubsystem = Object.prototype.hasOwnProperty.call(validSubSystems, subsystemStr);

    return hasDomain && hasSubsystem;
  }

  /**
   * Maps an RNC ID to its canonical filesystem location.
   */
  public static suggestPath(artifactId: string): string {
    const match = this.RNC_PATTERN.exec(artifactId);
    if (!match) {
      throw new Error(`Invalid RNC ID for path mapping: ${artifactId}`);
    }

    const domain = match[1];
    const subsystem = match[2];

    // Domain-to-Root Mapping
    const rootMap: Record<string, string> = {
      GVRN: "_governance",
      CORE: "axion-core",
      LAB: "nova_forge",
      SELT: "_governance/templates",
      COMM: "03_Avatar",
    };

    const root = rootMap[domain] || "unknown";

    // Subsystem-to-Folder Mapping
    const folderMap: Record<string, string> = {
      AVATAR: "03_Avatar",
      REGISTRY: "01_Registries",
      ENGINE: "forge",
      LEARNING: "06_Learning",
    };

    const folder = folderMap[subsystem] || subsystem.toLowerCase();

    return path.join(root, folder, `${artifactId}.md`);
  }

  /**
   * Applies a programmatic transformation with state-buffering to prevent truncation.
   */
  public static safeTransform(filePath: string, transformerFunc: (content: string) => string): void {
    if (!fs.existsSync(filePath)) {
      throw new Error(`Cannot transform non-existent file: ${filePath}`);
    }

    // 1. Read entire file into buffer (Kinetic -> Mind)
    const content = fs.readFileSync(filePath, "utf8");

    // 2. Apply transformation in-memory
    const transformedContent = transformerFunc(content);

    // 3. Validation: Ensure we haven't nuked the file (Zero Entropy protection)
    if (content.length > 0 && transformedContent.length === 0) {
      throw new Error(
        `Safety Trigger: Transformation returned 0 bytes for ${filePath}. Aborting write.`
      );
    }

    // 4. Atomic Write (Mind -> Substrate)
    fs.writeFileSync(filePath, transformedContent, "utf8");
    PhoenixLogger.info(`RNCEngine: Safely transformed ${filePath}`);
  }

  /**
   * Audits a directory for RNC compliance and returns a dissonance report.
   */
  public static syncFolder(folderPath: string): Array<{ file: string; issue: string; detectedId: string }> {
    const report: Array<{ file: string; issue: string; detectedId: string }> = [];

    const walk = (dir: string): void => {
      if (!fs.existsSync(dir)) {
        return;
      }
      const list = fs.readdirSync(dir);
      for (const file of list) {
        const fullPath = path.join(dir, file);
        const stat = fs.statSync(fullPath);
        if (stat.isDirectory()) {
          // Avoid scanning node_modules or system folders
          if (file !== "node_modules" && file !== ".git" && file !== "__pycache__") {
            walk(fullPath);
          }
        } else if (file.endsWith(".md")) {
          const artifactId = path.basename(file, ".md");
          if (!this.validateId(artifactId)) {
            report.push({
              file: fullPath,
              issue: "Non-compliant RNC ID",
              detectedId: artifactId,
            });
          }
        }
      }
    };

    walk(folderPath);
    return report;
  }
}
